domain/unitstate/state/state.go

```diff
@@ -177,25 +177,129 @@ func (st *State) SetUnitStateRelation(ctx domain.AtomicContext, uuid string, sta
-			return fmt.Errorf("deleting unit relation state: %w", err)
+			return errors.Errorf("deleting unit relation state: %w", err)
 		}
 
 		if err := tx.Query(ctx, iStmt, keyVals).Run(); err != nil {
-			return fmt.Errorf("setting unit relation state: %w", err)
+			return errors.Errorf("setting unit relation state: %w", err)
 		}
 		return nil
 	})
 }
+
+// GetUnitState returns the full agent state of the unit.
+// If no unit with the uuid exists, a [errors.UnitNotFound] error is returned.
+// If the units state is empty [errors.EmptyUnitState] error is returned.
+func (st *State) GetUnitState(ctx context.Context, uuid string) (unitstate.AgentState, error) {
+	db, err := st.DB()
+	if err != nil {
+		return unitstate.AgentState{}, errors.Errorf("getting db: %w", err)
+	}
+
+	id := unitUUID{UUID: uuid}
+	var name unitName
+	q := "SELECT &unitName.name FROM unit WHERE uuid = $unitUUID.uuid"
+	unitNameStmt, err := st.Prepare(q, name, id)
+	if err != nil {
+		return unitstate.AgentState{}, errors.Errorf("preparing select unit name statement: %w", err)
+	}
+
+	var state unitState
+	q = `
+	SELECT &unitState.*
+	FROM unit_state
+	WHERE unit_uuid = $unitUUID.uuid`
+	unitStateStmt, err := st.Prepare(q, state, id)
+	if err != nil {
+		return unitstate.AgentState{}, errors.Errorf("preparing select unit state statement: %w", err)
+	}
+
+	var charmKVs []unitCharmStateKeyVal
+	q = `
+	SELECT &unitCharmStateKeyVal.*
+	FROM unit_state_charm
+	WHERE unit_uuid = $unitUUID.uuid`
+	charmStateStmt, err := st.Prepare(q, unitCharmStateKeyVal{}, id)
+	if err != nil {
+		return unitstate.AgentState{}, errors.Errorf("preparing select unit charm state statement: %w", err)
+	}
+
+	var relationKVs []unitRelationStateKeyVal
+	q = `
+	SELECT &unitRelationStateKeyVal.*
+	FROM unit_state_relation
+	WHERE unit_uuid = $unitUUID.uuid`
+	relationStateStmt, err := st.Prepare(q, unitRelationStateKeyVal{}, id)
+	if err != nil {
+		return unitstate.AgentState{}, errors.Errorf("preparing select unit relation state statement: %w", err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, unitNameStmt, id).Get(&name)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return uniterrors.UnitNotFound
+		} else if err != nil {
+			return errors.Errorf("getting unit name: %w", err)
+		}
+
+		err = tx.Query(ctx, unitStateStmt, id).Get(&state)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("getting unit state: %w", err)
+		}
+
+		err = tx.Query(ctx, charmStateStmt, id).GetAll(&charmKVs)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("getting unit charm state: %w", err)
+		}
+
+		err = tx.Query(ctx, relationStateStmt, id).GetAll(&relationKVs)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("getting unit relation state: %w", err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return unitstate.AgentState{}, err
+	}
+
+	agentState := unitstate.AgentState{
```

Comment:
I would actually make another type for this to separate saved and unsaved snapshots.

The one that you query has no pointers. That way, you don't do a dance to make everything into a pointer, just to maybe dereference when packing RPC returns.